File: taskflow-backend/src/main/java/ma/rania/taskflow/service/TaskService.java
package ma.rania.taskflow.service;

import lombok.RequiredArgsConstructor;
import ma.rania.taskflow.dto.TaskDTO;
import ma.rania.taskflow.dto.TaskResponse;
import ma.rania.taskflow.model.Project;
import ma.rania.taskflow.model.Task;
import ma.rania.taskflow.model.User;
import ma.rania.taskflow.repository.ProjectRepository;
import ma.rania.taskflow.repository.TaskRepository;
import ma.rania.taskflow.repository.UserRepository;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Transactional
public class TaskService {

    private final TaskRepository    taskRepository;
    private final ProjectRepository projectRepository;
    private final UserRepository    userRepository;

    @Transactional(readOnly = true)
    public List<TaskResponse> getByProject(Long projectId) {
        return taskRepository.findByProjectId(projectId)
                .stream().map(TaskResponse::from).toList();
    }

    @Transactional(readOnly = true)
    public List<TaskResponse> getByProjectAndStatus(Long projectId, Task.Status status) {
        return taskRepository.findByProjectIdAndStatus(projectId, status)
                .stream().map(TaskResponse::from).toList();
    }

    @Transactional(readOnly = true)
    public List<TaskResponse> getMyTasks(Long userId) {
        return taskRepository.findByAssigneeId(userId)
                .stream().map(TaskResponse::from).toList();
    }

    @Transactional(readOnly = true)
    public TaskResponse getById(Long id) {
        return TaskResponse.from(findTask(id));
    }

    public TaskResponse create(TaskDTO dto) {
        Task task = new Task();
        applyDto(dto, task);
        return TaskResponse.from(taskRepository.save(task));
    }

    public TaskResponse update(Long id, TaskDTO dto) {
        Task task = findTask(id);
        applyDto(dto, task);
        return TaskResponse.from(taskRepository.save(task));
    }

    public TaskResponse updateStatus(Long id, Map<String, String> body) {
        Task task = findTask(id);
        String status = body.get("status");
        if (status == null) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Champ 'status' manquant");
        task.setStatus(Task.Status.valueOf(status));
        return TaskResponse.from(taskRepository.save(task));
    }

    public void delete(Long id) {
        taskRepository.delete(findTask(id));
    }

    // ── helpers ───────────────────────────────────────────────────────────────

    private Task findTask(Long id) {
        return taskRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Tâche introuvable : " + id));
    }

    private void applyDto(TaskDTO dto, Task task) {
        Project project = projectRepository.findById(dto.getProjectId())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Projet introuvable : " + dto.getProjectId()));

        task.setTitle(dto.getTitle());
        task.setDescription(dto.getDescription());
        task.setStatus(dto.getStatus());
        task.setPriority(dto.getPriority());
        task.setDueDate(dto.getDueDate());
        task.setProject(project);

        if (dto.getAssigneeId() != null) {
            User assignee = userRepository.findById(dto.getAssigneeId())
                    .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Utilisateur introuvable : " + dto.getAssigneeId()));
            task.setAssignee(assignee);
        } else {
            task.setAssignee(null);
        }
    }
}

File: taskflow-frontend/src/app/core/services/task.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Task, TaskDTO, TaskStatus } from '../models/task.model';
import { environment } from '../../../environments/environment';

@Injectable({ providedIn: 'root' })
export class TaskService {

  private readonly url = `${environment.apiUrl}/tasks`;

  constructor(private http: HttpClient) {}

  getByProject(projectId: number): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.url}/project/${projectId}`);
  }

  getByProjectAndStatus(projectId: number, status: TaskStatus): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.url}/project/${projectId}/status/${status}`);
  }

  getMyTasks(userId: number): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.url}/my-tasks/${userId}`);
  }

  getById(id: number): Observable<Task> {
    return this.http.get<Task>(`${this.url}/${id}`);
  }

  create(dto: TaskDTO): Observable<Task> {
    return this.http.post<Task>(this.url, dto);
  }

  update(id: number, dto: TaskDTO): Observable<Task> {
    return this.http.put<Task>(`${this.url}/${id}`, dto);
  }

  updateStatus(id: number, status: TaskStatus): Observable<Task> {
    return this.http.patch<Task>(`${this.url}/${id}/status`, { status });
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.url}/${id}`);
  }
}
